### ISSUE TITLE
Could not connect ZAP in remote API

### ISSUE THREAD

**Issue by dsbonafe** [zaproxy/zaproxy#5383]
**Describe the bug**

The request is locked when try to execute the method `scan` of `api.spider`.

**To Reproduce**
1 - Create a remote Linux machine visible from localhost. We're supposing it is on `10.152.165.2`.
2 - `ping 10.152.165.2` to grant the connectivity.
3 - In this machine, execute the following *docker* command to install zap in a container and run it.
```
docker run -u zap -p 8080:8080 -i owasp/zap2docker-stable zap.sh -daemon -host 0.0.0.0 -port 8080 -config api.addrs.addr.name=.* -config api.addrs.addr.regex=true -config api.disablekey=true           
```

4 - From localhost, execute the curl command:
```
curl -X GET "http://10.157.152.184:8080/xml/spider/action/scan/?url=https%3A%2F%2Fteperamenttest--selestrim.repl.co&"
```

5 - In your localhost (a different machine from which ZAP is installed), implement the following Snippet class:
```java
public class SimpleExample {
             //The next line contains the data where ZAP is installed and running.
             private static final String ZAP_ADDRESS = "10.152.165.2";
             private static final int ZAP_PORT = 8080;
             private static final String ZAP_API_KEY = null; // My api was configured 

             private static final String TARGET = "https://teperamenttest--selestrim.repl.co";

             public static void main(String[] args) {
                          ClientApi api = new ClientApi(ZAP_ADDRESS, ZAP_PORT, ZAP_API_KEY);

                          try {
                                       System.out.println("Spider : " + TARGET);
                                       ApiResponse resp = api.spider.scan(TARGET, null, null, null, null);
                          } catch (Exception e) {
                                       System.out.println("Exception : " + e.getMessage());
                                       e.printStackTrace();
                          }
             }
}

```
6 - Run it.

**Expected behavior**
We expect the scanner starting.

** Screenshot **
Not allowed.

**Software versions**
 - ZAP on the server: 2.6
 - ZAP Client Api: 1.6.0
 - OS: Windows 10
 - Java: 1.8.0_171
 - Browser: irrelevant

**Errors from the zap.log file**
Any exception was thrown.

**Additional context**
In the green line, we are creating the api as expected without any issues. We are calling the following constructor as expected:

public ClientApi(String zapAddress, int zapPort, String apiKey, boolean debug) {
                          proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(zapAddress, zapPort));
                          this.debug = debug;
                          this.zapAddress = zapAddress;
                          this.zapPort = zapPort;
                          this.apiKey = apiKey;
             }

Where zapAddres is 10.0.0.13 and port 8080.

So, we continue debugging until the red line. In this line, we are calling `scan()` method which calls ` api.callApi("spider", "action", "scan", map);`, which by their time calls `buildZapRequest()` method, which has the following line:

sb.append("http://zap/");

Which defines zap hardcoded as the name of solved IP of zapAddres.

Question 1: Where in the code "http://zap/" is being proxy to http://10.0.0.13:8080?

Once HttpRequest was built, the api open the connection with zap by the proxy and connect with `uc.connect()`.
Therefore, the next line should test the connection, and here is where the code stay locked forever:

```java
(...)
if (uc.getResponseCode() >= HttpURLConnection.HTTP_BAD_REQUEST) {
                                       return uc.getErrorStream();
                          }
```

This is the ClientApi.java@Line:359, in the method `getConnectionInputStream(HttpRequest request)`.
In this method, the generated request url in HttpRequest is: http://zap/xml/spider/action/scan/?url=https%3A%2F%2Fteperamenttest--selestrim.repl.co&

If I run the following command:
```
 curl -X GET "http://10.157.152.184:8080/xml/spider/action/scan/?url=https%3A%2F%2Fteperamenttest--selestrim.repl.co&" 
```

We obtain as answer the expected:
```html
 <?xml version="1.0" encoding="UTF-8" standalone="no"?><scan>0</scan> 
```
But, running by the ZAP Api the code stay locked as you can see here in the following values:
When in line `ClientApi.java@L361`:
```java
=> if (uc.getResponseCode() >= HttpURLConnection.HTTP_BAD_REQUEST) {
            return uc.getErrorStream();
        }
```
We have `request -> requestUri -> host` = `zap`. The value which was set in the `buildZapRequest` method.

And then, I run the line is locking the code… exactly when I try to execute the command `uc.getResponseCode()`.

And here are the values on proxy…

`proxy -> sa -> holder -> addr -> holder -> address == 168496141`
`proxy -> sa -> hostname == null`
`proxy -> sa -> port ==8080`
 
Could you help us on that?

Here is a thread I was talking about it with Kingthorin, the co-lead.

https://github.com/zaproxy/zaproxy/issues/4750

**Comment by kingthorin** [comment]
Why are you curl’ing against 184 when the docker is on 2?

When the docker image is up and running what do you get when you simply browse (without proxying) to localhost:8080 on the host system?

**Comment by dsbonafe** [comment]
Sorry. I've copied the wrong curl line. Fixed.

**Comment by kingthorin** [comment]
9k, let us know about direct browsing of the api web interface.

**Comment by dsbonafe** [comment]
It works as expected too.

**Comment by kingthorin** [comment]
Ok if you are proxying can you access it as http(s)://zap?

Can you browse it (http://ip:8080) remotely from the API client host?

Have you accessed the URL via ZAP before trying to spider?

**Comment by dsbonafe** [comment]
> Ok if you are proxying can you access it as http(s)://zap?

**Ans.**: No. I can't.

> Can you browse it (http://ip:8080) remotely from the API client host?

**Ans.:** Yes, I can. From my localhost, a machine different which one zap is running.

> Have you accessed the URL via ZAP before trying to spider?

**Ans.:** Yes, I can access ZAP API UI from remote, and can run curl commands from remote too.

**Comment by kingthorin** [comment]
> Have you accessed the URL via ZAP before trying to spider?
> Ans.: Yes, I can access ZAP API UI from remote, and can run curl commands from remote too.

Sorry I meant the target URL. (Ex has something been proxied or have you used the accessUrl method before trying to launch the spider?)

**Comment by dsbonafe** [comment]
Yes. But the target is not important too. I can access the target directly, using UI versions or curl.
I try to use `accessUrl()` method too, but I faced the same error.

**Comment by kingthorin** [comment]
Nope my concern was more that without having accessed something and having populated the Sites Tree (which is ZAP's understanding of the target) with at least a single entry then there's be nothing to spider (which usually leads to different and obvious errors...but still best to be sure).

**Comment by dsbonafe** [comment]
Yes. It is. But it's still not the case. hehe

**Comment by kingthorin** [comment]
@psiinon @thc202 any further ideas or suggestions?

**Comment by psiinon** [comment]
Is there anything in the zap.log file? 
https://github.com/zaproxy/zaproxy/wiki/FAQhelp#check-the-log-file

**Comment by dsbonafe** [comment]
Not at all. Nothing in logs.

Em sex, 7 de jun de 2019 04:13, Simon Bennetts <notifications@github.com>
escreveu:

**Comment by psiinon** [comment]
That narrows it down - it means that its a networking problem, ZAP is just not getting the requests.

**Comment by dsbonafe** [comment]
But other requests is catched by the ZAP. Just this one and few others, wasn't. If is this a network problem, should not be rational to think that ZAP could not receive any other request? But requests like createNewContext or getVersion or generateHTMLReport are catched by ZAP. My solution is being do not use your client, but implement a new client with REST. Using this approach for the same network, I got satisfatory results. But using your api client implementation the same workflow did not work. If it was a network problem, should it happen?

Thanks.

**Comment by thc202** [comment]
Have you checked if the Java process is not being blocked? (Windows likes to do that.)

If you want to replicate what the API client is using with `curl` you should use `http://zap/` while proxying through the address/port (`-x`). If that works then the client should work too.

**Comment by dsbonafe** [comment]
The program is running on linux and `http://zap/` did not work, what means that it's not being proxyed, I think. Should I run the docker command differently? Do you understand exactly what I have described in the issue report? The architecture is described there.
